// react-map-gl/dist/overlays/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _canvas = require('./canvas.react');

Object.defineProperty(exports, 'CanvasOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_canvas).default;
  }
});
Object.defineProperty(exports, 'ChoroplethOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_canvas).default;
  }
});
Object.defineProperty(exports, 'DraggablePointsOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_canvas).default;
  }
});
Object.defineProperty(exports, 'HTMLOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_canvas).default;
  }
});
Object.defineProperty(exports, 'ScatterplotOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_canvas).default;
  }
});
Object.defineProperty(exports, 'SVGOverlay', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_canvas).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
